### PR TITLE
Replaces ZipInputStream with ZipFile to fix Zip Slip vulnerability

### PR DESCRIPTION
**Pull request by DarshitChanpura**
More details about the vulenrability can be found on https://github.com/snyk/zip-slip-vulnerability
### Description
ZipInputStream can cause a potential [zipslip vulnerability](https://github.com/snyk/zip-slip-vulnerability). This PR aims at fixing the vulnerability by replacing it with ZipFile class.

### Issues Resolved
-

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java [hunk 1]
```diff
@@ -749,12 +752,11 @@ private Path unzip(Path zip, Path pluginsDir) throws IOException, UserException
                 if (entry.isDirectory() == false) {
                     try (OutputStream out = Files.newOutputStream(targetFile)) {
                         int len;
-                        while ((len = zipInput.read(buffer)) >= 0) {
+                        while ((len = zipFile.getInputStream(entry).read(buffer)) >= 0) {
```
Comment (dblock) [inline distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:755]: I think this reopens the stream every time we loop (or relies on the fact that `zipFile.getInputStream` memoizes a stream), let's do it once?

File: CHANGELOG.md [hunk 2]
```diff
@@ -79,6 +79,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Fix compression support for h2c protocol ([#4944](https://github.com/opensearch-project/OpenSearch/pull/4944))
 - Support OpenSSL Provider with default Netty allocator ([#5460](https://github.com/opensearch-project/OpenSearch/pull/5460))
 - Avoid negative memory result in IndicesQueryCache stats calculation ([#6917](https://github.com/opensearch-project/OpenSearch/pull/6917))
+- Replaces ZipInputStream with ZipFile to fix ZipSlip vulnerability ([#7230](https://github.com/opensearch-project/OpenSearch/pull/7230))
```
Comment (dblock) [inline CHANGELOG.md:82]: replace "fix ZipSlip..." by "fix CVE-..."

**Review by dblock — CHANGES_REQUESTED** [review]
Looks good if you can please make gradle check pass + address my comment on stream reuse and changelog entry. Thanks!

File: CHANGELOG.md (hunk 2 above)
Comment (DarshitChanpura) [reply CHANGELOG.md:82]: I don't know of a specific CVE for this vulnerability. But I do see a group of different vulnerabilities related to zip slip exploit. Refer here for more details:
- https://github.com/snyk/zip-slip-vulnerability
- https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=directory+traversal+crafted+zip

**Comment by DarshitChanpura** [comment]
Need to re-run  Gradle Precommit. [The failure](https://github.com/opensearch-project/OpenSearch/actions/runs/4748623226/jobs/8435026609?pr=7230#step:4:41) seems to be flaky on the Github runner's end.

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java [hunk 3]
```diff
@@ -748,13 +751,13 @@ private Path unzip(Path zip, Path pluginsDir) throws IOException, UserException
                 }
                 if (entry.isDirectory() == false) {
                     try (OutputStream out = Files.newOutputStream(targetFile)) {
+                        InputStream input = zipFile.getInputStream(entry);
```
Comment (dblock) [inline distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:754]: Does the input stream not need to be closed, including on error?

File: CHANGELOG.md (hunk 2 above)
Comment (dblock) [reply CHANGELOG.md:78]: Ok, I don't see a CVE in the list on https://github.com/snyk/zip-slip-vulnerability either. 

I think the correct spelling is with a space though, isn't it? "Zip Slip".

File: distribution/tools/plugin-cli/build.gradle [hunk 4]
```diff
@@ -78,3 +80,28 @@ thirdPartyAudit.ignoreViolations(
   'org.bouncycastle.jcajce.provider.ProvSunTLSKDF$TLSExtendedMasterSecretGenerator',
   'org.bouncycastle.jcajce.provider.ProvSunTLSKDF$TLSExtendedMasterSecretGenerator$2'
 )
+
+thirdPartyAudit.ignoreMissingClasses(
```
Comment (dblock) [inline distribution/tools/plugin-cli/build.gradle:84]: Why is this needed?

Comment (DarshitChanpura) [reply distribution/tools/plugin-cli/build.gradle:84]: Without these, the task `':distribution:tools:plugin-cli:thirdPartyAudit'` fails. See details [here](https://build.ci.opensearch.org/job/gradle-check/14349/consoleFull).

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java (hunk 3 above)
Comment (DarshitChanpura) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:754]: Yes, I will add `input.close()`.

**Comment by DarshitChanpura** [comment]
Gradle precommit failure seems to be a memory issue with GIthub runner. Re-run should probably solve it.

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java (hunk 3 above)
Comment (andrross) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:754]: Include it in the try-with-resources:

```
try (OutputStream out = Files.newOutputStream(targetFile);
    InputStream input = zipFile.getInputStream(entry)) {
...
}
```

File: distribution/tools/plugin-cli/build.gradle (hunk 4 above)
Comment (reta) [reply distribution/tools/plugin-cli/build.gradle:84]: > @reta do you know if this is ok?

Yeah, I think this is fine, the common-compress may have optional dependencies on ZSTD and other codecs, so we need to exclude these classes (we don't use these codecs here)

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java [hunk 5]
```diff
@@ -748,13 +751,14 @@ private Path unzip(Path zip, Path pluginsDir) throws IOException, UserException
...
```
Comment (reta) [inline distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:754]: :+1: to @dblock  and @andrross  and please use `zipInput.transferTo(out);`

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java [hunk 6]
```diff
@@ -716,10 +717,12 @@ private Path unzip(Path zip, Path pluginsDir) throws IOException, UserException
...
 
-        try (ZipInputStream zipInput = new ZipInputStream(Files.newInputStream(zip))) {
-            ZipEntry entry;
+        try (ZipFile zipFile = new ZipFile(zip, "UTF8", true, false)) {
+            final Enumeration<? extends ZipArchiveEntry> entries = zipFile.getEntries();
+            ZipArchiveEntry entry;
             byte[] buffer = new byte[8192];
-            while ((entry = zipInput.getNextEntry()) != null) {
+            while (entries.hasMoreElements()) {
+                entry = entries.nextElement();
```
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:725]: `entries.nextElement()` returns NoSucElementException when no more elements exists.
According to definition:
```
Throws:
NoSuchElementException – if no more elements exist.
```

**Review by reta — APPROVED** [review]

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
@DarshitChanpura want to manually backport to 2.x?

**Pull request merged**